src/client/hud.cpp

```diff
@@ -349,34 +346,56 @@ void Hud::drawLuaElements(const v3s16 &camera_offset)
 					inv, e->item, e->dir);
 				break; }
 			case HUD_ELEM_WAYPOINT: {
-				v3f p_pos = player->getPosition() / BS;
-				v3f w_pos = e->world_pos * BS;
-				float distance = std::floor(10 * p_pos.getDistanceFrom(e->world_pos)) /
-					10.0f;
-				scene::ICameraSceneNode* camera =
-					RenderingEngine::get_scene_manager()->getActiveCamera();
-				w_pos -= intToFloat(camera_offset, BS);
-				core::matrix4 trans = camera->getProjectionMatrix();
-				trans *= camera->getViewMatrix();
-				f32 transformed_pos[4] = { w_pos.X, w_pos.Y, w_pos.Z, 1.0f };
-				trans.multiplyWith1x4Matrix(transformed_pos);
-				if (transformed_pos[3] < 0)
+				if (calculateScreenPos(camera_offset, e, &pos))
 					break;
-				f32 zDiv = transformed_pos[3] == 0.0f ? 1.0f :
-					core::reciprocal(transformed_pos[3]);
-				pos.X = m_screensize.X * (0.5 * transformed_pos[0] * zDiv + 0.5);
-				pos.Y = m_screensize.Y * (0.5 - transformed_pos[1] * zDiv * 0.5);
+				v3f p_pos = player->getPosition() / BS;
+				pos += v2s32(e->offset.X, e->offset.Y);
 				video::SColor color(255, (e->number >> 16) & 0xFF,
 										 (e->number >> 8)  & 0xFF,
 										 (e->number >> 0)  & 0xFF);
-				core::rect<s32> size(0, 0, 200, 2 * text_height);
 				std::wstring text = unescape_translate(utf8_to_wide(e->name));
-				font->draw(text.c_str(), size + pos, color);
-				std::ostringstream os;
-				os << distance << e->text;
-				text = unescape_translate(utf8_to_wide(os.str()));
-				pos.Y += text_height;
-				font->draw(text.c_str(), size + pos, color);
+				// waypoints reuse the item field to store precision, item = precision + 1
+				u32 item = e->item;
+				float precision = item == 0 ? 10.0f : (item - 1);
+				bool draw_precision = precision > 0;
```

Comment:
(adding a space between the decoding and the using makes it easier to read)

```suggestion
				bool draw_precision = precision > 0;

```